x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go

```diff
@@ -0,0 +1,194 @@
+	s.App.StakeibcKeeper.IBCKeeper.ClientKeeper.SetClientState(s.Ctx, "client-0", state)
+	s.App.StakeibcKeeper.IBCKeeper.ConnectionKeeper.SetConnection(s.Ctx, connectionId, connectionEnd)
+	s.App.StakeibcKeeper.IBCKeeper.ConnectionKeeper.SetConnection(s.Ctx, connectionId, connectionEnd)
+
+	s.App.ICAControllerKeeper.RegisterInterchainAccount(s.Ctx, connectionId, owner)
+	s.App.ICAControllerKeeper.SetActiveChannelID(s.Ctx, connectionId, portId, "channel-0")
+	channel, _ := s.App.StakeibcKeeper.IBCKeeper.ChannelKeeper.GetChannel(s.Ctx, "connection-0", "icacontroller-STRIDE.REDEMPTION")
+	channel.State = channeltypes.OPEN
+	channel.Counterparty = channeltypes.NewCounterparty(portId, "channel-0")
+	s.App.StakeibcKeeper.IBCKeeper.ChannelKeeper.SetChannel(s.Ctx, portId, "channel-0", channel)
+
+	msgSend := banktypes.MsgSend{
+		FromAddress: redemptionAccount.Address,
+		ToAddress:   receiverAddr,
+		Amount:      redemptionAmount,
+	}
+
+	return ClaimUndelegatedTestCase{
+		redemptionRecordId: redemptionRecordId,
+		validMsg: stakeibc.MsgClaimUndelegatedTokens{
+			Creator:    senderAddr,
+			HostZoneId: "GAIA",
+			Epoch:      1,
+			Sender:     senderAddr,
+		},
+		expectedTxMsg: stakeibckeeper.IcaTx{
+			ConnectionId: connectionId,
+			Msgs:         []sdk.Msg{&msgSend},
+			Account:      redemptionAccount,
+			Timeout:      uint64(types.DefaultICATimeoutNanos),
+		},
+		redemptionRecord: redemptionRecord,
+		hostZone:         hostZone,
+	}
+}
+
+func (s *KeeperTestSuite) TestClaimUndelegatedTokensSuccessful() {
+	tc := s.SetupClaimUndelegatedTokens()
+
+	userRedemptionRecord, err := s.App.StakeibcKeeper.GetClaimableRedemptionRecord(s.Ctx, &tc.validMsg)
+	s.Require().NoError(err, "get redemptions record should not error")
+
+	actualTxMsg, err := s.App.StakeibcKeeper.GetRedemptionTransferMsg(s.Ctx, userRedemptionRecord, tc.validMsg.HostZoneId)
+	s.Require().NoError(err, "get redemption transfer msg should not error")
+	s.Require().Equal(tc.expectedTxMsg, *actualTxMsg, "redemption transfer message")
+
+	//Confirm the redemption record has been flagged as not claimable
+	s.App.StakeibcKeeper.FlagRedemptionRecordsAsClaimed(s.Ctx, userRedemptionRecord, 1)
+	redemptionRecord, found := s.App.RecordsKeeper.GetUserRedemptionRecord(s.Ctx, tc.redemptionRecordId)
+	s.Require().True(found)
+	s.Require().False(redemptionRecord.IsClaimable)
+
+	// Confirm pending claims added
+	pendingClaims := s.App.StakeibcKeeper.GetAllPendingClaims(s.Ctx)
+	s.Require().Equal(1, len(pendingClaims))
+	pendingRedemptionRecordIds := pendingClaims[0].UserRedemptionRecordIds
+	s.Require().Equal(1, len(pendingRedemptionRecordIds))
+	s.Require().Equal(tc.redemptionRecordId, pendingRedemptionRecordIds[0])
+	// QUESTION: Anything else to check that I'm missing?
+}
+
+func (s *KeeperTestSuite) TestClaimUndelegatedTokensNoUserRedemptionRecord() {
```

Comment:
Should we add a failure test with bad inputs to submittxs? You know better than me whether that should be tested here or elsewhere

Reply:
Yep agreed